### ISSUE TITLE
ghc-9.4+/Cabal-3.8+ continue to build package after source module files fail with errors

### ISSUE THREAD

**Issue by juhp** [haskell/cabal#10243]
**Describe the bug**
This seems like a noticeable change of behavior (or arguably even a regression perhaps?).

While updating Fedora Haskell to ghc-9.6, I noticed that a couple of packages seemed
to fail to build silently. On further investigation of the build.logs (for darcs and hadolint)
I realised they _were_ erroring _but_ the compilation error was happening much earlier
with the build continuing to build dozens or hundreds of further modules before giving up.
I certainly found this confusing, though I had not really noticed this under ghc 9.4.
My initial thought was that this must be a bug, but I guess it is fair to call it a significant change of behaviour.

NB these distro builds are not using cabal-install but Setup (or so-called Cabal v1).

TL;DR:
For Cabal < 3.8 (ghc < 9.4), building a package with an error would always stop compilation as early as possible:
ie right after a source module files failed to compile with one or more errors.

But with Cabal-3.8 and later (ghc 9.4+), building a package with errors now seems to continue compiling after source module files fail to compile, continuing to build until something needs one of the failed modules it appears.

**To Reproduce**

You can reproduce with this https://github.com/juhp/test-Cabal-build toy project.

Try building it with ghc-9.4/9.6 and 9.2 with each of their Cabal libraries.

**Expected behavior**
I was expecting compilation to stop after the first error.

**System information**
 - Operating system: Fedora Linux
 - `Cabal` 3.10.3 , `ghc` 9.6.6

**Additional context**
A smaller real package example I found later was cheapskate.

During discussion on Matrix, Artem suggested there should at least be a flag to control this.

**Comment by juhp** [comment]
Hmm I can also reproduce with stack and `ghc --make` closing this for now, doesn't seem to be Cabal indeed.

**Comment by ulysses4ever** [comment]
Right. Cabal basically uses `ghc --make`.
